src/collapse/dtrace.rs

```diff
@@ -109,21 +151,152 @@ impl Collapse for Folder {
 
         None
     }
-}
 
-impl From<Options> for Folder {
-    fn from(opt: Options) -> Self {
-        Self {
-            stack: VecDeque::default(),
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            cache_inlines: Vec::new(),
-            opt,
-            stack_str_size: 0,
-        }
+    #[cfg(test)]
+    fn set_nstacks_per_job(&mut self, n: usize) {
+        self.opt.nstacks_per_job = n;
+    }
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize) {
+        self.opt.nthreads = n;
     }
 }
 
 impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        assert!(self.occurrences.is_concurrent());
+        assert!(self.opt.nstacks_per_job > 0);
+        assert!(self.opt.nthreads > 1);
+
+        enum Message {
+            Job(Vec<u8>),
+            Shutdown,
+        }
+
+        crossbeam::thread::scope(|scope| {
+            // Spin up the threadpool / worker threads.
+            let (tx_input, rx_input) = channel::unbounded();
+            let (tx_output, rx_output) = channel::unbounded();
+            let mut handles = Vec::with_capacity(self.opt.nthreads);
+            for _ in 0..self.opt.nthreads {
+                let rx_input = rx_input.clone();
+                let tx_output = tx_output.clone();
+
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        cache_inlines: Vec::default(),
+                        occurrences,
+                        stack: VecDeque::default(),
+                        stack_str_size: 0,
+                        opt,
+                    };
+                    loop {
+                        // Receive the data...
+                        let data = match rx_input.recv().unwrap() {
```

Comment:
I _think_ this loop can just be:
```rust
while let Message::Job(data) = rx.input.recv().unwrap()`
```
or `s/Message::Job/Some/` if you replace `Message` with `Option`.

Reply:
Unfortunately when we got rid of the `Message` type and started sending just an `Option<Vec<u8>>`, the poor compiler seems to need a type annotation to figure out what the type of the data coming across is. In order to use `while let...` I'd have to put in an ugly type annotation in when creating the channel. How about we just go with the `match` statement, which still requires a type annotation now, but it's less ugly than doing it at the creation of the channel. What do you think?

Reply:
Ended up going with `while let...` as you can from the latest. Going to resolve this.